src/crypto/qat/qcccrypto.h

```diff
@@ -0,0 +1,174 @@
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Lesser General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2.1 of the License, or (at your option) any later version.
+ *
+ */
+
+#ifndef QCCCRYPTO_H
+#define QCCCRYPTO_H
+
+#include <atomic>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <pthread.h>
+extern "C" {
+#include "cpa.h"
+#include "lac/cpa_cy_sym.h"
+#include "lac/cpa_cy_im.h"
+#include "qae_mem.h"
+#include "icp_sal_user.h"
+#include "icp_sal_poll.h"
+#include "qae_mem_utils.h"
+}
+
+class QccCrypto {
+
+  public:
+    CpaCySymCipherDirection qccOpType;
+
+    QccCrypto();
+    ~QccCrypto();
+
+    int perform_op(unsigned char* out, const unsigned char* in, size_t size,
+        uint8_t *iv,
+        uint8_t *key,
+        CpaCySymCipherDirection op_type);
+
+  private:
+
+    static const size_t AES_256_IV_LEN = 16;
+    static const size_t AES_256_KEY_SIZE = 32;
+    static const size_t QCC_MAX_RETRIES = 100;
+
+    /*
+     * Struct to hold an instance of QAT to handle the crypto operations. These
+     * will be identified at the start and held until the destructor is called
+     * To-Do: Ganesh
+     * The struct was creating assuming that we will use all the instances.
+     * Expand current implementation to allow multiple instances to operate
+     * independently.
+     */
+    struct QCCINST
+    {
+      CpaInstanceHandle *cyInstHandles;
+      Cpa16U numInstances;
+    } *qccInst;
+
+    /*
+     * QAT Crypto Session
+     * Crypto Session Context and setupdata holds
+     * priority, type of crypto operation (cipher/chained),
+     * cipher algorithm (AES, DES, etc),
+     * single crypto or multi-buffer crypto.
+    */
+    struct QCCSESS
+    {
+      CpaCySymSessionSetupData sessStpData;
+      Cpa32U sessCtxSz;
+      CpaCySymSessionCtx sessCtx;
+      std::atomic_bool in_use;
+    } *qccSess;
+
+    /*
+     * Cipher Memory Allocations
+     * Holds bufferlist, flatbuffer, cipher opration data and buffermeta needed
+     * by QAT to perform the operation. Also buffers for IV, SRC, DEST.
+    */
+    struct QCCOPMEM
+    {
+      // Op common  items
+      bool ismemalloc;
+      CpaCySymOpData *symopdata;
+      Cpa32U buffmetasize;
+      Cpa32U numbuffers;
+      Cpa32U buffsize;
+
+      //Src data items
+      Cpa8U *srcbuffmeta;
+      CpaBufferList *srcbufflist;
+      CpaFlatBuffer *srcbuffflat;
+      Cpa8U *srcbuff;
+      Cpa8U *ivbuff;
+    } *qccOpMem;
+
+    /*
+     * Function to find available instance to handle the current op
+    */
+    int qccGetFreeInstance();
+
+    /*
+     * Contiguous Memory Allocator and de-allocator. We are using the usdm
+     * driver that comes along with QAT to get us direct memory access using
+     * hugepages.
+     * To-Do: Ganesh
+     * Maybe a kernel based one for later?
+    */
+    static __inline void qccContigMemFree(void **ptr)
```

Comment:
might want to use `inline` instead of `__inline`, since we have the luxury to use C++'s `inline` specifier, not the vendor specific one for C.

Reply:
might want to name this method `qcc_contig_mem_free` to be more consistent with rest part of this project.